### PR TITLE
CURATOR-599 Configurable ZookeeperFactory by ZKClientConfig

### PR DESCRIPTION
**Pull request by liran2000**
Option to use ZooKeeper client config.

This seems mandatory for using zookeeper.request.timeout for preventing
the potential race condition of hanging indefinitely, as described at
the ticket.

### CODE REVIEW

**Comment by liran2000** [comment]
it depends on ZooKeeper 3.6.1+ which recently updated.

It is passing locally, but failing in the CI build in GitHub, can anyone please help with that ?

@eolivelli @cammckenzie @tisonkun 

Possibly it is since Maven running there without 'clean' step, but I am not sure.
(mvn install -DskipTests=true -Dmaven.javadoc.skip=true -B -V)
Failing on java.lang.NoSuchMethodError: org.apache.zookeeper.admin.ZooKeeperAdmin: method <init>(Ljava/lang/String;ILorg/apache/zookeeper/Watcher;ZLorg/apache/zookeeper/client/ZKClientConfig;)V not found

**Comment by eolivelli** [comment]
taking a look now

**Comment by eolivelli** [comment]
This is an error in the compatibility tests for 3.5.
Basically those tests are there in order to prove that even if you have only ZK 3.5 in the classpath the basic features of Curator work

image

**Comment by eolivelli** [comment]
Probably we have to detect the availability of that method and fallback to legacy code in case we are on an older version of ZK.

**Comment by eolivelli** [comment]
@cammckenzie @Randgalt do you mind taking a look as well?

**Review by eolivelli — APPROVED** [review]
LGTM

**Review by cammckenzie — APPROVED** [review]

**Review by cammckenzie — CHANGES_REQUESTED** [review]
Some minor comments, otherwise, LGTM

**Review by tisonkun — APPROVED** [review]
I second @cammckenzie 's suggestion and other parts LGTM.

File: curator-framework/src/main/java/org/apache/curator/framework/imps/CuratorFrameworkImpl.java
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/CuratorFrameworkImpl.java:237]: I'm -1 on using `instanceof`. Wouldn't it be better to add a `default` method to `ZookeeperFactory` that takes the `ZKClientConfig`? It could be optional and forward to the older versions when not present. The default version in `DefaultZookeeperFactory` would throw an exception.

**Comment by liran2000** [comment]
Thank you all for the reviews and info, I addressed the comments. Can you please have another review and merge it ?

@eolivelli @cammckenzie @tisonkun @Randgalt

File: curator-framework/src/main/java/org/apache/curator/framework/imps/CuratorFrameworkImpl.java
Comment (liran2000) [reply curator-framework/src/main/java/org/apache/curator/framework/imps/CuratorFrameworkImpl.java]: Understood. Done.

File: curator-framework/src/test/java/org/apache/curator/framework/imps/TestFramework.java
Comment (liran2000) [reply curator-framework/src/test/java/org/apache/curator/framework/imps/TestFramework.java]: Done

File: curator-framework/src/test/java/org/apache/curator/framework/imps/TestFramework.java
Comment (eolivelli) [inline curator-framework/src/test/java/org/apache/curator/framework/imps/TestFramework.java:78]: Usually loggers are static fields
`private static final Logger log = LoggerFactory.getLogger(TestFramework.class);`

Comment (liran2000) [reply curator-framework/src/test/java/org/apache/curator/framework/imps/TestFramework.java:78]: I wrote it exactly as already implemented at the same package, to be consistent:

https://github.com/apache/curator/blob/b9dc9bd4056dba18c29a4f3127df43b930449abe/curator-framework/src/test/java/org/apache/curator/framework/imps/TestFrameworkBackground.java#L56

https://github.com/apache/curator/blob/b9dc9bd4056dba18c29a4f3127df43b930449abe/curator-framework/src/test/java/org/apache/curator/framework/imps/TestFrameworkEdges.java#L78

Comment (eolivelli) [reply curator-framework/src/test/java/org/apache/curator/framework/imps/TestFramework.java:78]: sorry.
you are right.
I forgot that we are in Curator and we have different conventions.

thanks for checking out

**Review by eolivelli — APPROVED** [review]
LGTM

waiting for @cammckenzie final feedback

**Review by cammckenzie — APPROVED** [review]
There a couple of minor formatting issues, but we can fix these when we merge.

LGTM, thanks @liran2000

**Comment by liran2000** [comment]
@eolivelli thank you, can you please merge it and then resolve the Jira issue ?

File: curator-framework/src/test/java/org/apache/curator/framework/imps/TestFramework.java
Comment (eolivelli) [inline curator-framework/src/test/java/org/apache/curator/framework/imps/TestFramework.java:1107]: Can you please fix formatting here?

Otherwise I will do it as follow up.
I will merge this patch today.

We are also going to cut a release as soon as possible

Comment (liran2000) [reply curator-framework/src/test/java/org/apache/curator/framework/imps/TestFramework.java]: Sure, done.
BTW, at my IDE this formatting, formatting was fine, depends on IDE config, I will notice it.

Comment (liran2000) [reply curator-framework/src/test/java/org/apache/curator/framework/imps/TestFramework.java]: @eolivelli thanks, can be merged and then resolve the Jira issue

**Comment by eolivelli** [comment]
probably you are linking the wrong JIRA
https://issues.apache.org/jira/browse/CURATOR-599 "Hanging indefinitely on some scenarios since zookeeper.request.timeout cannot be configured"

**Comment by eolivelli** [comment]
I will be happy to commit once we are pointing to the right JIRA

**Comment by liran2000** [comment]
@eolivelli  it is the right JIRA issue. The issue should be resolved when zookeeper.request.timeout can be configured - with this commit, it can be configured, then the issue should not occur. It was Hanging indefinitely, and when ZooKeeper request timeout is configured - it will get a timeout as expected, and not hang indefinitely . See below line that added at the test to configure it, and more details are there in the JIRA itself. If something is not clear you can let me know.

https://github.com/apache/curator/pull/391/files#diff-b423c4cc97cc1282306109606a1ab1060ce818cd1b04d6ea7c7493ca657bbf5dR1095

zkClientConfig.setProperty(ZKClientConfig.ZOOKEEPER_REQUEST_TIMEOUT, zookeeperRequestTimeout);

https://issues.apache.org/jira/browse/CURATOR-599 "Hanging indefinitely on some scenarios since zookeeper.request.timeout cannot be configured"

**Comment by liran2000** [comment]
@eolivelli any other clarification needed for the comment above ?
when zookeeper.request.timeout can be configured and used, it can resolve the issue at Jira:
CURATOR-599 - Hanging indefinitely on some scenarios since zookeeper.request.timeout cannot be configured

**Comment by eolivelli** [comment]
committed to master branch

**Comment by liran2000** [comment]
Thank you!

**Pull request closed without merging**
